force-app/main/default/classes/RollupRecord.cls

```diff
@@ -124,25 +167,253 @@ public with sharing class RollupRecord {
+                }
+            }
+            // Last Service Date is only on Contact and Program Engagement no need to check
+            // when the field is not present
+            if (lastServiceDateField == null) {
+                continue;
+            }
+
+            if (delivery.isDelivered()) {
+                if (
+                    isNew && delivery.getDeliveryDate() > lastServiceDate ||
+                    lastServiceDate == null
+                ) {
+                    lastServiceDate = delivery.getDeliveryDate();
+                } else if (!isNew && delivery.getDeliveryDate() == lastServiceDate) {
+                    lastServiceDate = null;
+                    reprocessLastServiceDate = true;
+                }
+            }
+        }
+
+        if (reprocessLastServiceDate || reprocessConsecutiveAbsences) {
+            processExistingDeliveries(
+                reprocessLastServiceDate,
+                reprocessConsecutiveAbsences
+            );
+        }
+    }
+
+    private void processUpdates() {
+        Map<Id, ServiceDelivery> originalDeliveryById = new Map<Id, ServiceDelivery>();
+        for (ServiceDelivery originalDelivery : originalDeliveries) {
+            originalDeliveryById.put(originalDelivery.getId(), originalDelivery);
+        }
+        Boolean reprocessLastServiceDate = false;
+        Boolean reprocessConsecutiveAbsences = false;
+
+        for (ServiceDelivery delivery : deliveries) {
+            ServiceDelivery originalDelivery = originalDeliveryById.get(delivery.getId());
+
+            Boolean deliveryDateChanged =
+                delivery.getDeliveryDate() != originalDelivery.getDeliveryDate();
+
+            if (delivery.isAbsent() && !originalDelivery.isAbsent()) {
+                incrementAbsent();
+                reprocessConsecutiveAbsences = true;
+
+                if (originalDelivery.isPresent()) {
+                    decrementPresent();
+                }
+            }
+
+            if (delivery.isPresent() && !originalDelivery.isPresent()) {
+                incrementPresent();
+                reprocessConsecutiveAbsences = true;
+
+                if (originalDelivery.isAbsent()) {
+                    decrementAbsent();
+                }
+            }
+
+            if (
+                delivery.isDelivered() != originalDelivery.isDelivered() ||
+                deliveryDateChanged
+            ) {
+                reprocessLastServiceDate = true;
+            }
+        }
+
+        processExistingDeliveries(reprocessLastServiceDate, reprocessConsecutiveAbsences);
+    }
+
+    private void processExistingDeliveries(
+        Boolean reprocessLastServiceDate,
+        Boolean reprocessConsecutiveAbsences
+    ) {
+        if ((consecutiveAbsencesField == null && lastServiceDateField == null)) {
+            return;
+        }
+
+        reprocessConsecutiveAbsences =
+            reprocessConsecutiveAbsences && consecutiveAbsencesField != null;
```

Comment:
Another silly question here. Do we need the null check here since it will do an early return if both consecutiveAbsencesField and lastServiceDateField is null.

Reply:
One or the other vs both could be null. I know, this is another part of the code that feels icky. I'm null checking the fields in way to many places.